fix: Use extattr on FreeBSD instead of xattr

JunePrimavera opened the pull request:
As per issue #26 

`plpfuse` is unable to compile due to `xattr` missing on FreeBSD, this patch uses FreeBSD's replacement, `extattr` when compiling on such systems.

I'm currently unable to compile this, but should resolve the issue (Worst case I've made a typo somewhere and it fails to compile)

rrthomas commented:
Thanks very much for this! Minor nit: could you please use `#ifdef` rather than `#if defined(…)`, like the rest of the file? (I admit the code base is inconsistent across components, but I'd rather try to keep what harmony there is!)

JunePrimavera commented:
> Thanks very much for this! Minor nit: could you please use `#ifdef` rather than `#if defined(…)`, like the rest of the file? (I admit the code base is inconsistent across components, but I'd rather try to keep what harmony there is!)

I've now fixed this in the latest commit, should be all good now. Was able to test and verify that it compiles (at least on FreeBSD, haven't looked into NetBSD or compiled on linux yet)

> It's been a long time since I wrote this code, so I may be completely forgetting what it's doing and how it works, but it seems to be calling FreeBSD functions to get and set xattrs on files on the local filing system, which is surely the wrong way around? These xattrs are supposed to be being read and written on the EPOC filing system.
> 
> So in particular, I don't expect to see any calls to extattr APIs.

I can't speak as an expert here, but it's probably setting xattrs locally, since that's likely more documented than on the EPOC filing system (Avoiding having to potentially write your own software to get/set xattrs?). Not sure if this is actually the case, but it is what first comes to my mind looking at the code.

kapfab commented:
As I understand it, there is currently no get/set API call (just a check of the flags) but the change seems to replace some flag values (XATTR_*) by API calls.

rrthomas commented:
> I may be misunderstanding this - Surely my patch shouldn't be doing anything to the local filesystem any more than on non FreeBSD systems? If the EPOC filing system is being exposed to the host via FUSE, surely the system calls should still work as expected?

Exactly. It shouldn't! But it was, because it was making extattr API calls to change attributes of local files.

rrthomas commented:
> As I understand it, there is currently no get/set API call (just a check of the flags) but the change seems to replace some flag values (XATTR_*) by API calls.

`extattr_set/get_file` reads and sets extended attributes on FreeBSD. This is not what this code is trying to do, it's reading and setting attributes of EPOC files.

JunePrimavera commented:
This makes sense - Thank you.
Completely misunderstood what the code was trying to do, re-read through the docs and will try make my patch better suited for this.

It doesn't seem like `XATTR_CREATE` or `XATTR_REPLACE` have any exact drop-in replacement for FreeBSD (At least not within `extattr`)
Unfortunately this means that a more complex solution is required and it can't just be a single-line fix. I believe the functionality is there, though.

JunePrimavera commented:
After looking into this more - Is there any reason not to just check `flags` against what `XATTR_CREATE` and `XATTR_REPLACE`  *should* be if they were defined on FreeBSD.
Considering the code isn't actually reading or writing anything (Presumably that's handled elsewhere?) only checking the flags, as long as the flags passed into `plp_setxattr` are correct; the rest of the code compiles completely fine on FreeBSD, it's just `XATTR_CREATE` and `XATTR_REPLACE` which are missing - Which are just integers at the end of the day.

Granted, this highly depends on how `plpfuse` is interacting with the rest of the codebase and how `plp_setxattr` is being used and it's nowhere near as elegant as on other systems. Feels like a hacky solution, but might be plausible?

rrthomas commented:
> After looking into this more - Is there any reason not to just check `flags` against what `XATTR_CREATE` and `XATTR_REPLACE` _should_ be if they were defined on FreeBSD.

Indeed, that's what I was expecting!

> Granted, this highly depends on how `plp_setxattr` is being used and it's nowhere near as elegant as on other systems. Feels like a hacky solution, but might be plausible?

Context: EPOC has nothing like extended attributes in its file system (which is FAT), and we're just using extended attributes to be able to read and set the FAT attributes "archive", "hidden" and "system". We present these to the host OS as extended attributes under the name `user.epoc`.

I'm not sure what you're suggesting is a hack compared to other systems (I assume you mean the solution for FreeBSD is a hack?). I was hoping that FreeBSD would have constants equivalent to `XATTR_CREATE` (but as you can see, we only use that to give a sensible error) and `XATTR_REPLACE` (again, we just use that to give a sensible error message).

If these constants have no sensible value on FreeBSD, the relevant error returns can just be commented out, I guess, and the behaviour on FreeBSD will just be slightly less precise.

I'm surprised that the flag values don't have symbolic names on FreeBSD, is that really the case‽ If it is, then could you include a link to the relevant documentation, and give the magic numbers symbolic names at the top of the file (near `XATTR_NAME` and friends)?

Finally, your current PR seems to have accidentally reformatted the file; please could you fix that to remove the reformatting?

Thanks very much for persevering with this!

JunePrimavera commented:
> Finally, your current PR seems to have accidentally reformatted the file; please could you fix that to remove the reformatting?

Whoops, must've accidentally left auto formatting on

> I'm not sure what you're suggesting is a hack compared to other systems (I assume you mean the solution for FreeBSD is a hack?). I was hoping that FreeBSD would have constants equivalent to `XATTR_CREATE` (but as you can see, we only use that to give a sensible error) and `XATTR_REPLACE` (again, we just use that to give a sensible error message).

The solution for FreeBSD feels like a hack to me, since it's working on additional assumptions - Such as assuming the values for each flag never change across different systems. I feel this is a reasonable assumption to make as this is only being used to give a sensible error, if the assumption is wrong the rest of the software should work just fine regardless - Just me thinking too deeply about it, I usually try to avoid making assumptions about the system my own software is running on. 

> I'm surprised that the flag values don't have symbolic names on FreeBSD, is that really the case‽ If it is, then could you include a link to the relevant documentation, and give the magic numbers symbolic names at the top of the file (near `XATTR_NAME` and friends)?

The issue is that the flags don't really have an equivalent on FreeBSD to begin with, so there's nothing to give a symbolic name - `extattr` and `xattr` are just too different. To get around this, I've simply defined the flags on FreeBSD as what they would be on Linux, from the kernel source (https://github.com/torvalds/linux/blob/master/include/uapi/linux/xattr.h):

```c
#define XATTR_CREATE	0x1	/* set value, fail if attr already exists */
#define XATTR_REPLACE	0x2	/* set value, fail if attr does not exist */
```

This is why I feel it's a bit of a hacky solution - It makes the software compile, but assumes whatever flags are passed into `plp_setxattr` match (Which is dependent on how it's being used elsewhere).


Additional note - This would work on any system with no `xattr`, is it worth changing the check for FreeBSD to something more general? Ideally something like a flag or env variable, for other operating systems with the same issue.

rrthomas commented:
> The issue is that the flags don't really have an equivalent on FreeBSD to begin with, so there's nothing to give a symbolic name - `extattr` and `xattr` are just too different.

OK, so this is more of a FUSE issue. How are these FUSE endpoints invoked on FreeBSD; indeed, are they ever invoked? It's possible we should simply be commenting out the whole thing.

If there's no way for userland to invoke these endpoints with the same flag values we use, then the whole exercise is at best pointless.

> Additional note - This would work on any system with no `xattr`, is it worth changing the check for FreeBSD to something more general? Ideally something like a flag or env variable, for other operating systems with the same issue.

As above, I wonder whether it would actually work, or whether those endpoints would simply never be reached.

So, the next thing to do seems to be to investigate whether and how these endpoints work for FUSE on BSD.

JunePrimavera commented:
> OK, so this is more of a FUSE issue. How are these FUSE endpoints invoked on FreeBSD; indeed, are they ever invoked? It's possible we should simply be commenting out the whole thing.

Which endpoints in particular?
From what I can see, FUSE should work just fine on FreeBSD - Though it's only supported on a "best-effort" basis.
There are related tests within FreeBSD (https://cgit.freebsd.org/src/tree/tests/sys/fs/fusefs/xattr.cc?id=4b60fac3a5c7a53f6d1a4480f37eb906bbacf21b) which all seem to be using the expected behavior for FreeBSD, with the exception of one test where FUSE appears to use Linux behavior on FreeBSD for whatever reason (Though, I don't think that functionality is super relevant here)

```c
/*
 * On FreeBSD, if the user passes an insufficiently large buffer then the
 * filesystem is supposed to copy as much of the attribute's value as will fit.
 *
 * On Linux, however, the filesystem is supposed to return ERANGE.
 *
 * libfuse specifies the Linux behavior.  However, that's probably an error.
 * It would probably be correct for the filesystem to use platform-dependent
 * behavior.
 *
 * This test case covers a filesystem that uses the Linux behavior
 * TODO: require FreeBSD Behavior.
 */
TEST_F(Getxattr, erange)
{
	char data[10];
	uint64_t ino = 42;
	int ns = EXTATTR_NAMESPACE_USER;
	ssize_t r;

	expect_lookup(RELPATH, ino, S_IFREG | 0644, 0, 1);
	expect_getxattr(ino, "user.foo", ReturnErrno(ERANGE));

	r = extattr_get_file(FULLPATH, ns, "foo", data, sizeof(data));
	ASSERT_EQ(-1, r);
	ASSERT_EQ(ERANGE, errno);
}
```
With this in mind, I'm not entirely sure whether it should be commented out or not - It doesn't seem unlikely for it to be using Linux behavior on FreeBSD, even when it really shouldn't, but on the other hand, it does defeat the point of giving a user a reasonable error message if the behavior is inconsistent between different operations with FUSE.

rrthomas commented:
> Which endpoints in particular?

Specifically, `setxattr`, `getxattr`, `listxattr` and `removexattr`.

> /*
>  * On FreeBSD, if the user passes an insufficiently large buffer then the
>  * filesystem is supposed to copy as much of the attribute's value as will fit.
>  *
>  * On Linux, however, the filesystem is supposed to return ERANGE.
>  *
>  * libfuse specifies the Linux behavior.  However, that's probably an error.
>  * It would probably be correct for the filesystem to use platform-dependent
>  * behavior.
>  *
>  * This test case covers a filesystem that uses the Linux behavior
>  * TODO: require FreeBSD Behavior.
>  */
> TEST_F(Getxattr, erange)

Ah cool, so it seems these endpoints do work on FreeBSD. OK, so how are they called? What I mean by that is, for example, the `open` endpoint will be called when a BSD `open` syscall is made to a FUSE file system. On Linux, the `getxattr` endpoint will be reached from a `getxattr` syscall. On FreeBSD, there is no `getxattr` syscall as I understand it, so presumably the endpoint is called by some sort of impedance matching from the corresponding `extattr` call?

JunePrimavera commented:
Curiously, I can't find any mention of a workaround for `xattr` FreeBSD within FUSE.
Looking at their examples within the `libfuse` repository (https://github.com/libfuse/libfuse/blob/43e36ffed84f73528b67ca1835760bf5895c9d60/example/passthrough.c), if there's no `xattr` on the system, it simply doesn't import `sys/xattr.h`

It seems the implementation of FUSE within the FreeBSD kernel has been modified with this in mind (https://cgit.freebsd.org/src/tree/sys/fs/fuse/fuse_vnops.c) though it's unclear to me how exactly these endpoints are meant to be used.

```c
/*
 * List extended attributes
 *
 * The FUSE_LISTXATTR operation is based on Linux's listxattr(2) syscall, which
 * has a number of differences compared to its FreeBSD equivalent,
 * extattr_list_file:
 *
 * - FUSE_LISTXATTR returns all extended attributes across all namespaces,
 *   whereas listxattr(2) only returns attributes for a single namespace
 * - FUSE_LISTXATTR prepends each attribute name with "namespace."
 * - If the provided buffer is not large enough to hold the result,
 *   FUSE_LISTXATTR should return ERANGE, whereas listxattr is expected to
 *   return as many results as will fit.
 */
/*
    struct vop_listextattr_args {
	struct vop_generic_args a_gen;
	struct vnode *a_vp;
	int a_attrnamespace;
	struct uio *a_uio;
	size_t *a_size;
	struct ucred *a_cred;
	struct thread *a_td;
    };
*/
```

There's no mention of flags anywhere, though - Which makes me think the flag is not necessary for FreeBSD (And as such can be commented out).
Though, I am really not sure if I am interpreting this correctly (Or even looking at the right code)

rrthomas commented:
> It seems the implementation of FUSE within the FreeBSD kernel has been modified with this in mind (https://cgit.freebsd.org/src/tree/sys/fs/fuse/fuse_vnops.c) though it's unclear to me how exactly these endpoints are meant to be used.
> 
> There's no mention of flags anywhere, though - Which makes me think the flag is not necessary for FreeBSD (And as such can be commented out). Though, I am really not sure if I am interpreting this correctly (Or even looking at the right code)

Many thanks for digging into this. The FreeBSD approach seems to be "let's just support what is easy to support, it might make more sense in future".

So, I suggest we do the same, make one harmless assumption, that the missing constants are `#ifdef`d when they are defined, and put `#ifdef` guards around their use. This will mean that if anyone tries to create or replace an xattr on FreeBSD, they won't get errors when they should, or they'll get a misleading error. If that ever bothers anyone, maybe they can help us get documented constants in FreeBSD.

JunePrimavera commented:
> So, I suggest we do the same, make one harmless assumption, that the missing constants are `#ifdef`d when they are defined, and put `#ifdef` guards around their use. This will mean that if anyone tries to create or replace an xattr on FreeBSD, they won't get errors when they should, or they'll get a misleading error. If that ever bothers anyone, maybe they can help us get documented constants in FreeBSD.

So - comment out the line entirely, rather than manually define the constants?

rrthomas commented:
Yup, I don't think it makes sense for plptools to unilaterally define them (even by copying Linux). If the whole point of plpfuse were extended attributes, maybe, but in fact they're just a hack; it's not even obvious to me they're the right hack, but they're the hack we have. (As a side note, I think there's a FAT FUSE FS? If so, what does it do for the extra FAT attributes? But as I can't be bothered to look right now, please don't anyone else feel they have to!)

rrthomas commented:
This is looking good; please when it's done can you squash it down to a single commit and then I'll probably make some tedious nit-picky requests.

JunePrimavera commented:
I've squashed it down to a single commit 

One last thing to note is that this introduces a new warning (Though doesn't prevent it from compiling) since I've removed the offending if statements on FreeBSD and this *could* lead to the function not returning anything on FreeBSD. Not sure if it's best to leave it like that, or just return the result of the statement assuming it's true (or false), which would at least suppress the warning, but might lead to additional unintended behavior.

JunePrimavera commented:
> > Is there anything more that needs to be done to this other than test? Also, I guess this wouldn't need much of a change to make it work with NetBSD?
> 
> There's just my two outstanding comments on the changes.

To my memory I've implemented everything - which additional changes you are looking for?

rrthomas commented:
Please see my pending review comments and reply as appropriate. Maybe I simply haven't spotted changes you already made!

JunePrimavera commented:
> Please see my pending review comments and reply as appropriate. Maybe I simply haven't spotted changes you already made! 

I think I'm missing which comments you're referring to, there's probably things I need to change, just something I'm overlooking!

rrthomas commented:
Haha, GitHub is complicated! Look at the Files Changed tab of this PR, scroll down, and you should find them. (Any GitHub experts, if there was something else I should have done to make them (more) visible, let me know!)

jbmorley commented:
Did you commit your review? I don’t see the comments clearly on my end…On 5 Jul 2025, at 10:04, Reuben Thomas ***@***.***> wrote:﻿rrthomas left a comment (plptools/plptools#27)
Haha, GitHub is complicated! Look at the Files Changed tab of this PR, scroll down, and you should find them. (Any GitHub experts, if there was something else I should have done to make them (more) visible, let me know!)

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

rrthomas commented:
> Did you commit your review? I don’t see the comments clearly on my end.

Thanks, now done that.

JunePrimavera commented on plpfuse/fuse.c:
```diff
@@ -18,11 +18,20 @@
+
+  https://github.com/plptools/plptools/pull/27
+*/
+#elif HAVE_ATTR_XATTR_H
```
We don't have the header file on FreeBSD (Since the functionality is supposed to be replaced by `extattr`), so it can't be included (At least, not without some trickery).

rrthomas replied:
Sure, so on FreeBSD the symbol `HAVE_ATTR_XATTR_H` will not be defined. No need for any special case.

rrthomas commented on plpfuse/fuse.c:
```diff
@@ -473,10 +486,15 @@ static int plp_setxattr(const char *path, const char *name, const char *value, s
     debuglog("setxattr succeeded");
     return 0;
   } else {
+    #ifdef __FreeBSD__
```
Please use `#ifdef XATTR_REPLACE` (& similarly for `XATTR_CREATE`) rather than `#ifdef __FreeBSD__`. More general, and it's a safe bet that `XATTR_{REPLACE,CREATE} are preprocessor defines.

JunePrimavera replied:
I have now implemented this

rrthomas commented on plpfuse/fuse.c:
```diff
@@ -18,11 +18,12 @@
 #include <sys/time.h>
 #include <syslog.h>
-#ifdef HAVE_ATTR_XATTR_H
+#if HAVE_ATTR_XATTR_H
```
Please leave this as `#ifdef`. It's undefined if we don't have the header.

rrthomas commented on plpfuse/fuse.c:
```diff
@@ -473,10 +476,13 @@ static int plp_setxattr(const char *path, const char *name, const char *value, s
     else
+    #endif
       return -ENOTSUP;
+    
```
Please remove this stray extra blank line.

rrthomas commented:
Great stuff, please can you squash the commits into one?

rrthomas commented:
> @rrthomas I’ve noticed you asking for this on other PRs. I’m curious if there’s a reason we don’t just get GitHub to perform a squash commit on merge?

No, not really, just caution, as I haven't used that much. Thanks for the nudge, I will try to remember to try it in future.

jbmorley commented:
> > @rrthomas I’ve noticed you asking for this on other PRs. I’m curious if there’s a reason we don’t just get GitHub to perform a squash commit on merge?
> 
> No, not really, just caution, as I haven't used that much. Thanks for the nudge, I will try to remember to try it in future.

Cool! I have found it to work well on my other projects and it definitely eases the workflow. I typically change the default commit message it proposes to match the PR title and description as you then have a clear way for the PR author to set them and it’s just a one-click to accept them.

rrthomas commented:
Note: I rewrote the commit message after merging (and force-pushed), but didn't touch the actual changes.

Thanks all for getting this over the line. I'm sorry it was such a drawn-out process for what in the end was a tiny change. Of course part of it is a Cunning Plan to get others involved in maintenance!

rrthomas commented:
> @rrthomas I think perhaps in the future we should refrain from force pushing to main. I know it's not a widely used project at this point, but I think it's important to avoid history rewriting on main for downstream consumers.

I agree that practice will have to stop! I don't think it was a big deal in this case, as I did it within a couple of minutes of the merge; at this stage I'd be more worried about annoying other developers than users. But it was laziness; I think what I probably should have done was checked out the PR branch and, assuming it was set up to allow it, rewritten that.

rrthomas commented:
Thanks for that, should be a good discipline.
